### ISSUE TITLE
Sessions are still being regenerated on every page

### ISSUE THREAD

**Issue by ericnicolaas** [wpninjas/ninja-forms#2579]
This is a long-standing issue. See #954, #1502, #1236, #1649.

I previously submitted a PR which fixes this: https://github.com/wpninjas/ninja-forms/issues/1803 

The PR and all issues are closed, but this is still not fixed. 

You can reproduce the problem quite easily: 
```
add_action( 'init', function() {
    if ( class_exists('WP_Session') ) {
        echo WP_Session::get_instance()->session_id;
    }    
});
```
Reload the page a few times and you'll see the session ID is different on every page load. That effectively makes the session entirely pointless. 

Worse, you're using the `WP_Session` library, which is used by other plugins too. The way `WP_Session` works is that every installed plugin that uses it will share the same session. From your own code: 

```
if ( ! defined( 'WP_SESSION_COOKIE' ) ) {
   define( 'WP_SESSION_COOKIE', 'nf_wp_session' );
}
```

https://github.com/wpninjas/ninja-forms/blob/master/includes/Session.php#L46-L48

You'll see that this is the exact same way that other plugins who use `WP_Session` are defining the constant: 

- [Easy Digital Downloads](https://github.com/easydigitaldownloads/easy-digital-downloads/blob/master/includes/class-edd-session.php#L84-L86)
- [Charitable](https://github.com/Charitable/Charitable/blob/master/includes/public/class-charitable-session.php#L49-L51)
- [Give](https://github.com/WordImpress/Give/blob/master/includes/class-give-session.php#L96-L99)

In other words, Ninja Forms, Charitable, Easy Digital Downloads, Give and just about anyone else who uses `WP_Session` in their plugin is going to share the same session. 

The reason your implementation of `WP_Session` is breaking things is because you are _not_ setting the cookie on page load: 

<img width="520" alt="screenshot_2016-09-15_12_53_58" src="https://cloud.githubusercontent.com/assets/676582/18537309/bb4ac15a-7b43-11e6-917c-50abb9e42b35.png">

As a result, until some action happens to take place that involves Ninja Forms setting something on the session, Ni

**Comment by kjohnson** [comment]
Thanks @ericnicolaas . I'm looking into this now.

**Comment by kjohnson** [comment]
Looks like we are only using this for the Submissions Controller, specifically the Resume functionality for add-ons such as PayPal that require intermediary page redirects.

**Comment by ericnicolaas** [comment]
@kjohnson Is there any reason why those require the session to _not_ be set on page load?

**Comment by nickyoung87** [comment]
Is this going to be fixed in a release? I am working on a plugin and need data from WP Session to persist. When Ninja Forms is enabled it runs into trouble, but if I uncomment the `$this->set_cookie()` mentioned above then everything works fine. Just curious if it will be added or not so I know what direction I need to take.

**Comment by nickyoung87** [comment]
So yeah it's definitely the issue from that issue on NF GH. 

Changing Line 115 of ninja-forms/includes/Libraries/Session/class-wp-session.php to uncommented fixes the problem.

Here is the line pointing at their GH code: https://github.com/wpninjas/ninja-forms/blob/master/includes/Libraries/Session/class-wp-session.php#L115

I commented on their issue to see if they plan on getting it into a release. I am not sure the approach from here. Should we figure out a way to workaround it or point users to change that line if they must have NF enabled with ours?

**Comment by pderksen** [comment]
@kstover @kjohnson Thanks for looking into this. @nickyoung87 is the main developer on WP Simple Pay Pro 3 which is nearly out of beta and currently it's incompatible for the reason he mentioned.

If you'd like a copy and license for our plugin to check compatibility and code let me know or shoot me an email (phil AT wpsimplepay.com). Thanks!

**Comment by kstover** [comment]
@kjohnson - Can you give me an update on the status of this issue?

@pderksen - That would be awesome.

@nickyoung87 - Thanks for your input. Hopefully we can squash this stupid thing once and for all.

**Comment by kjohnson** [comment]
@kstover I'm looking into it as a part of this sprint.

**Comment by kjohnson** [comment]
@kstover I just opened #2792.

/cc @pderksen, @nickyoung87, @ericnicolaas

**Comment by nickyoung87** [comment]
@kjohnson If it helps. We use it to display payment details (like the amount, item name, etc) after a page redirect. So that is why when it gets cleared it messes with our stuff, because the data gets cleared out and our redirect page will tell us we have an error because that data no longer exists.

**Comment by kstover** [comment]
@nickyoung87 Does that solve the problem? Do we need to dig into the issue more?

**Comment by nickyoung87** [comment]
@kstover Seems to have done the trick!

**Comment by ericnicolaas** [comment]
@kjohnson @kstover Done the trick with Charitable too :)

**Comment by kstover** [comment]
Merged the fix into develop branch. :)
